cmd/crank/xpkg/install.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package xpkg
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/alecthomas/kong"
+	"github.com/google/go-containerregistry/pkg/name"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	v1 "github.com/crossplane/crossplane/apis/pkg/v1"
+	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
+	"github.com/crossplane/crossplane/internal/version"
+	"github.com/crossplane/crossplane/internal/xpkg"
+
+	// Load all the auth plugins for the cloud providers.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+const (
+	errPkgIdentifier = "invalid package image identifier"
+	errKubeConfig    = "failed to get kubeconfig"
+	errKubeClient    = "failed to create kube client"
+)
+
+// InstallCmd is exported so that it can be re-used by the beta xpkg subcommand.
+
+// InstallCmd installs a package.
+type InstallCmd struct {
+	Kind string `arg:"" help:"Kind of package to install. One of \"provider\", \"configuration\", or \"function\"." enum:"provider,configuration,function"`
```

Comment:
There is no YAML. This command essentially builds one (or rather, the internal struct representation) from its arguments.

Reply:
I am probably overlooking a technical detail but what I'm getting as is what prevents us from reading the yaml stream and loading the `Kind`s inside instead of requiring the user to specify the package type?

Reply:
Oh gotcha. Yes that would technically be possible. It's an interesting idea, but I lean toward leaving this as is. The CLI would need to pull down the package and introspect it. It doesn't do that (anywhere) today. We'd presumably have to tackle proxies, caching, and all that good stuff to make it work.

Despite largely rewriting the implementation of this command it's a pretty small diff on the previous implementation functionally. We go from `crossplane install provider` (where provider is a subcommand) to `crossplane xpkg install provider` (where provider is an argument).